File: pkg/masipam/ipam.go
package masipam

import (
	"context"
	"os"
	"path/filepath"

	"github.com/containernetworking/cni/pkg/invoke"
	"github.com/containernetworking/cni/pkg/types"
)

// ExecAdd calls the IPAM plugin to allocate an IP
// ExecAdd 调用外部 IPAM 插件 (如 host-local) 来申请 IP 地址。
// CNI 的委托机制允许我们复用现有的 IP 管理能力，而不需要自己造轮子。
func ExecAdd(ipamType string, netconf []byte) (types.Result, error) {
	// 2. 执行委托 (Delegate)
	// 本质上是 fork/exec 执行那个二进制文件，并把我们的 Stdin (配置) 和环境变传给它
	// 注意：CNI >= 0.8.0 的 DelegateAdd 第一个参数接受的是 *plugin name* (例如 "host-local")
	// 而不是绝对路径。invoke 库内部会自己再找一次。
	return invoke.DelegateAdd(context.Background(), ipamType, netconf, nil)
}

// ExecDel 调用外部 IPAM 插件释放 IP 地址。
func ExecDel(ipamType string, netconf []byte) error {
	// 委托执行 DEL
	return invoke.DelegateDel(context.Background(), ipamType, netconf, nil)
}

// locations 返回 CNI 插件的查找路径
func locations() []string {
	// 优先检查环境变量 CNI_PATH (多个路径用冒号分隔)
	// 默认路径为 /opt/cni/bin (这是所有 Kubernetes 节点的标准路径)
	paths := filepath.SplitList(os.Getenv("CNI_PATH"))
	if len(paths) == 0 {
		paths = []string{"/opt/cni/bin"}
	}
	return paths
}

File: src/ipam.rs
use anyhow::{Context, Result};
use std::io::Write;
use std::process::{Command, Stdio};

pub fn exec_add(ipam_type: &str, stdin_data: &[u8]) -> Result<String> {
    delegate("ADD", ipam_type, stdin_data)
}

pub fn exec_del(ipam_type: &str, stdin_data: &[u8]) -> Result<String> {
    delegate("DEL", ipam_type, stdin_data)
}

pub fn exec_check(ipam_type: &str, stdin_data: &[u8]) -> Result<String> {
    delegate("CHECK", ipam_type, stdin_data)
}

fn delegate(command: &str, ipam_type: &str, stdin_data: &[u8]) -> Result<String> {
    // 1. Find plugin path (simplified: look in /opt/cni/bin)
    let plugin_path = format!("/opt/cni/bin/{}", ipam_type);

    // 2. Prepare command
    // 这里等价于在 Shell 中执行以下命令（假设 ipam_type 是 "host-local"）：
    // export CNI_COMMAND=ADD (或者 DEL/CHECK)
    // export CNI_PATH=/opt/cni/bin
    // echo '{ "cniVersion": "...", "ipam": { ... } }' | /opt/cni/bin/host-local
    //
    // 关键点：
    // 1. 通过环境变量传参 (CNI_COMMAND)。
    // 2. 通过 Stdin 管道传入 JSON 配置 (child.stdin)。
    // 3. 捕获 Stdout 作为返回值 (child.stdout)，里面包含分配到的 IP。
    let mut child = Command::new(&plugin_path)
        .env("CNI_COMMAND", command)
        .env("CNI_PATH", "/opt/cni/bin") // Standard path
        // Pass through other CNI env vars ideally, but for MVP minimal set
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .with_context(|| format!("Failed to spawn IPAM plugin {}", plugin_path))?;

    // 3. Write config to stdin
    if let Some(mut stdin) = child.stdin.take() {
        stdin.write_all(stdin_data)?;
    }

    // 4. Wait for output
    let output = child.wait_with_output()?;

    if !output.status.success() {
        return Err(anyhow::anyhow!("IPAM plugin failed: {:?}", output.status));
    }

    let result = String::from_utf8(output.stdout)?;
    Ok(result)
}
